main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1074,6 +1074,27 @@ object Namer {
         case e => NamedAst.Expression.Par(e, loc)
       }
 
+    case WeededAst.Expression.ParYield(frags, exp, loc) =>
+      // Visit patterns and collect an env
+      val (pats, finalEnv) = frags.foldRight((List.empty[NamedAst.Pattern], env0)) {
+        case (WeededAst.ParYield.Fragment(pat, _, _), (pats, envAcc)) =>
+          val (p, env1) = visitPattern(pat, uenv0)
+          (p :: pats, envAcc ++ env1)
+      }
+
+      // Fragments exps are visited in env0 and paired with the previously visited pattern
+      val fragsVal = traverse(frags.zip(pats)) {
```

Comment:
I wonder how it would look just to traverse everything once and to have a `var finalEnv` that gets updated. What do you thing?

Reply:
I guess we could do that. That saves us a fold. I think I'm just allergic to mutability :joy:

Reply:
I think `traverse` does some parallel stuff. If we have mutable var we would need a mutex around it, yeah?
```scala
  /**
    * A fast implementation of traverse.
    */
  private def fastTraverse[T, S, E](xs: Iterable[T])(f: T => Validation[S, E]): Validation[List[S], E] = {
    // Check if the sequence is empty.
    if (xs.isEmpty)
      return Validation.SuccessNil

    // Two mutable arrays to hold the intermediate results.
    val successValues = mutable.ArrayBuffer.empty[S]
    val failureStream = mutable.ArrayBuffer.empty[LazyList[E]]

    // Apply f to each element and collect the results.
    for (x <- xs) {
      f(x) match {
        case Success(v) => successValues += v
        case Failure(e) => failureStream += e
      }
    }

    // Check whether we were successful or not.
    if (failureStream.isEmpty) {
      Success(successValues.toList)
    } else {
      Failure(failureStream.foldLeft(LazyList.empty[E])(_ #::: _))
    }
  }
```

Reply:
It doesn't look like this is parallel. Just a for-loop, no?

Reply:
Ahh yes, I totally had it mixed up with something else... oops :eyes: